### ISSUE TITLE
redis: Log spam warning when redis is loading the dataset in memory

### ISSUE THREAD

**Issue by ijsnow** [sourcegraph/sourcegraph#2904]
Some concerning repo-updater logs were reported from a customer:

```
22:08:33       repo-updater | t=2019-03-06T22:08:33+0000 lvl=warn msg="failed to execute redis command" cmd=SETEX error="LOADING Redis is loading the dataset in memory"
22:08:33       repo-updater | t=2019-03-06T22:08:33+0000 lvl=warn msg="failed to execute redis command" cmd=SETEX error="LOADING Redis is loading the dataset in memory"
22:08:33       repo-updater | t=2019-03-06T22:08:33+0000 lvl=warn msg="failed to execute redis command" cmd=SETEX error="LOADING Redis is loading the dataset in memory"
```

These logs didn't appear to correlate with any broken functionality but we should be aware and investigate eventually. They were noticed while looking into the cause of another issue and ultimately took up a non-zero amount of time that could have been used focusing on the issue at hand.

**Comment by tsenart** [comment]
Upstream issue: https://github.com/antirez/redis/issues/4624

**Comment by slimsag** [comment]
@tsenart it looks like you linked the upstream issue, closed this issue, but there wasn't any resolution here? Or was something not linked?

A customer (https://app.hubspot.com/contacts/2762526/company/554338610/) is now facing this same error in their server logs, and it looks quite bad. From what I can tell, in both the customer's logs and in the original issue here this indicates a real issue where we are not properly waiting for Redis to start up fully -- something we should definitely invest in fixing.

Note that this happens in many more cases than just `repo-updater`, e.g. on startup:

```
frontend |  !!!!!  Redis is required
frontend | Problem: Redis is unavailable or misconfigured
frontend | Error: LOADING Redis is loading the dataset in memory
frontend | Possible fix: Start a Redis server listening at port 127.0.0.1:6379
frontend | Skip this check by setting the env var SRC_SKIP_REQS="Redis" (separate
frontend | multiple entries with spaces). Note: Sourcegraph may not function
frontend | properly without Redis.
frontend | t=2020-01-21T19:46:24+0000 lvl=eror msg="System requirement checks failed (see above for more information)." failed=[Redis]
frontend | t=... lvl=eror msg="Unable to delete old cache data in redis search. Please report this issue." error="LOADING Redis is loading the dataset in memory"
```

And this does appear to be a fatal issue when encountered, it stops the server from starting: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8fa196dd64e6998a12a6b4520773363f786ca5b5/-/blob/cmd/frontend/internal/cli/serve_cmd.go#L152

**Comment by tsenart** [comment]
@slimsag: I closed this because there were no recent reports of this behaviour. Thanks for re-opening, given a customer is now facing this again. Is this customer running the server image? If so, this might be alleviated by #3300, which @unknwon will be working on this iteration.

But in any case, despite us making start up faster, I think we still need to ensure we wait for Redis to be ready before starting Sourcegraph processes.

Might be enough to use this: https://github.com/Hronom/wait-for-redis

I'm taking this on soon.

**Comment by keegancsmith** [comment]
@tsenart the bit of logs complaining about redis implements a wait for redis in Go. It is simply just polling redis with a ping. We already have infrastructure to do this, we likely just are not doing it in repo-updater. Its our `sysreq` pkg. https://github.com/sourcegraph/sourcegraph/blob/28424bd97f782f61000121e8eb1da3d441f5f23c/internal/redispool/sysreq.go#L13

**Comment by tsenart** [comment]
Thanks for the context @keegancsmith. I think it'd be more robust to do it at startup time outside of Go, or we risk further instances of this problem as new uses of Redis by unsuspecting engineers pop up. WDYT?

**Comment by keegancsmith** [comment]
the pattern for sysreq is if you import a package it registers it's sysreq in a global pattern via init. Then you wait for your sysreq's to load. The issue with waiting for DB/redis to start before starting everything else is that slows startup down. Personally I think more uses of sysreq makes sense.

**Comment by tsenart** [comment]
@keegancsmith: As per @slimsag's pasted logs, our `sysreq` package is actually working as it shuts down the server because Redis isn't available. What we want to do, however, is to wait for Redis, not simply quit if it's not available at that point.

**Comment by keegancsmith** [comment]
If that is possible, we should increase how long we wait for before giving up.

**Comment by tsenart** [comment]
Fixed by #8114
